libs/deer/src/impls/core/ops.rs

```diff
@@ -0,0 +1,403 @@
+
+    fn expecting(&self) -> Document {
+        Self::Value::reflection()
+    }
+
+    fn visit_value<D>(
+        self,
+        discriminant: Self::Discriminant,
+        deserializer: D,
+    ) -> Result<Self::Value, VisitorError>
+    where
+        D: Deserializer<'de>,
+    {
+        match discriminant {
+            BoundDiscriminant::Unbounded => deserializer
+                .deserialize_optional(UnitVariantVisitor)
+                .map(|_| Bound::Unbounded)
+                .attach(Location::Variant("Unbounded"))
+                .change_context(VisitorError),
+            BoundDiscriminant::Included => T::deserialize(deserializer)
+                .map(Bound::Included)
+                .attach(Location::Variant("Included"))
+                .change_context(VisitorError),
+            BoundDiscriminant::Excluded => T::deserialize(deserializer)
+                .map(Bound::Excluded)
+                .attach(Location::Variant("Excluded"))
+                .change_context(VisitorError),
+        }
+    }
+}
+
+pub struct BoundReflection<T: ?Sized>(fn() -> *const T);
+
+impl<T> Reflection for BoundReflection<T>
+where
+    T: Reflection + ?Sized,
+{
+    fn schema(doc: &mut Document) -> Schema {
+        #[derive(serde::Serialize)]
+        enum BoundOneOf {
+            Included(Reference),
+            Excluded(Reference),
+            Unbounded(Reference),
+        }
+
+        // TODO: the case where "Unbounded" as a single value is possible cannot be
+        //  represented right now with deer Schema capabilities
+        Schema::new("object").with("oneOf", [
+            BoundOneOf::Included(doc.add::<T>()),
+            BoundOneOf::Excluded(doc.add::<T>()),
+            BoundOneOf::Unbounded(doc.add::<<() as Deserialize>::Reflection>()),
+        ])
+    }
+}
+
+impl<'de, T> Deserialize<'de> for Bound<T>
+where
+    T: Deserialize<'de>,
+{
+    type Reflection = BoundReflection<T::Reflection>;
+
+    fn deserialize<D: Deserializer<'de>>(de: D) -> Result<Self, DeserializeError> {
+        de.deserialize_enum(BoundEnumVisitor(PhantomData))
+            .change_context(DeserializeError)
+    }
+}
+
+identifier! {
+    enum RangeIdent {
+        Start = "start" | b"start" | 0,
+        End = "end" | b"end" | 1,
+    }
+}
+
+struct RangeFieldVisitor<'a, T, U> {
+    start: &'a mut Option<T>,
+    end: &'a mut Option<U>,
+}
+
+impl<'a, 'de, T, U> FieldVisitor<'de> for RangeFieldVisitor<'a, T, U>
+where
+    T: Deserialize<'de>,
+    U: Deserialize<'de>,
+{
+    type Key = RangeIdent;
+    type Value = ();
+
+    fn visit_value<D>(self, key: Self::Key, deserializer: D) -> Result<Self::Value, VisitorError>
+    where
+        D: Deserializer<'de>,
+    {
+        match key {
+            RangeIdent::Start => {
+                let value = T::deserialize(deserializer)
+                    .attach(Location::Field("start"))
+                    .change_context(VisitorError)?;
+
+                if self.start.replace(value).is_some() {
```

Comment:
This took me a second to understand. If the `start` can be replaced, returning a `some` it means it was already set?

Reply:
fixed in [cd00879](https://github.com/hashintel/hash/pull/2422/commits/cd008794257d4363c4b1d4a5986273e5c7b17a16)